sql/src/main/java/io/crate/planner/operators/LogicalPlanner.java

```diff
@@ -0,0 +1,178 @@
+import io.crate.analyze.WhereClause;
+import io.crate.analyze.relations.QueriedRelation;
+import io.crate.analyze.symbol.FieldsVisitor;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.RefVisitor;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.planner.MultiPhasePlan;
+import io.crate.planner.Plan;
+import io.crate.planner.Planner;
+import io.crate.planner.SubqueryPlanner;
+import io.crate.planner.projection.builder.ProjectionBuilder;
+import io.crate.planner.projection.builder.SplitPoints;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+/**
+ * Planner which can create a {@link Plan} using intermediate {@link LogicalPlan} nodes.
+ */
+public class LogicalPlanner {
+
+    public static final int NO_LIMIT = -1;
+
+    public Plan plan(QueriedRelation queriedRelation,
+                     Planner.Context plannerContext,
+                     ProjectionBuilder projectionBuilder) {
+        LogicalPlan logicalPlan = plan(queriedRelation)
+            .build(extractColumns(queriedRelation.querySpec().outputs()))
+            .tryCollapse();
+
+        Plan plan = logicalPlan.build(
+            plannerContext,
+            projectionBuilder,
+            LogicalPlanner.NO_LIMIT,
+            0,
+            null
+        );
+        SubqueryPlanner subqueryPlanner = new SubqueryPlanner(plannerContext);
+        return MultiPhasePlan.createIfNeeded(plan, subqueryPlanner.planSubQueries(queriedRelation.querySpec()));
+    }
+
+    static LogicalPlan.Builder plan(QueriedRelation relation) {
+        SplitPoints splitPoints = SplitPoints.create(relation.querySpec());
+        return fetchOrEval(
+            limit(
+                orderBy(
+                    filter(
+                        groupByOrAggregate(
+                            whereAndCollect(relation, splitPoints.toCollect(), relation.where()),
+                            relation.groupBy(),
+                            splitPoints.aggregates()),
+                        relation.having()
+                    ),
+                    relation.orderBy()
+                ),
+                relation.limit(),
+                relation.offset()
+            ),
+            relation.querySpec().outputs()
+        );
+    }
+
+    private static LogicalPlan.Builder fetchOrEval(LogicalPlan.Builder source, List<Symbol> outputs) {
+        return usedColumns -> new FetchOrEval(source.build(Collections.emptySet()), usedColumns, outputs);
+    }
+
+    private static LogicalPlan.Builder limit(LogicalPlan.Builder source, @Nullable Symbol limit, @Nullable Symbol offset) {
+        if (limit == null && offset == null) {
+            return source;
+        }
+        return usedColumns -> new Limit(
```

Comment:
I like these functional defs.

Reply:
I was/am btw also thinking about moving these methods into the LogicalPlan classes. E.g have some kind of `create` factory methods. Thoughts?